### PR TITLE
[Routing] Fix PSR-4 directory loader for abstract classes

### LINKED ISSUE #48168: [Routing] Psr4DirectoryLoader issue with abstract classes
### Symfony version(s) affected

6.2-beta

### Description

The new `Psr4DirectoryLoader` only supports `attribute` or `annotation` type so it will use the `AnnotationClassLoader` which don’t want to deal with abstract classes: https://github.com/symfony/symfony/blob/b02a68911aded41f7b4929c0834004ebefe6976f/src/Symfony/Component/Routing/Loader/AnnotationClassLoader.php#L114-L116
(See https://github.com/symfony/symfony/commit/946ccb6bd287c46bccf2f4f94a0ddded33e1c0c6 for rationale.)

If `AnnotationDirectoryLoader` [skips them](https://github.com/symfony/symfony/blob/b02a68911aded41f7b4929c0834004ebefe6976f/src/Symfony/Component/Routing/Loader/AnnotationDirectoryLoader.php#L56-L58), `Psr4DirectoryLoader` does not meaning it will crash on abstract classes.

### How to reproduce

Configure your routes to use the PSR-4 route loader, eg.

```php
<?php

use Symfony\Component\Routing\Loader\Configurator\RoutingConfigurator;

return static function (RoutingConfigurator $routes)
{
    $routes->import(['path' => '../src/Controller', 'namespace' => 'App\Controller'], 'annotation');
};
```

and make one of your controllers extends an abstract one (it does not need to define any method).

Refresh the cache to see the error.

### Possible Solution

`Psr4DirectoryLoader` should skip abstract classes. As it means it will need to create an instance of `ReflectionClass` it could be useful to make `AnnotationClassLoader::load` accept these.

### Additional Context

Spotted from https://github.com/symfony/symfony/discussions/48161

### PR DESCRIPTION
**Pull request by derrabus**
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #48168, Replaces #48169
| License       | MIT
| Doc PR        | N/A

### CODE REVIEW

File: src/Symfony/Component/Routing/Loader/Psr4DirectoryLoader.php [hunk 1]
```diff
@@ -83,7 +83,7 @@ function (\SplFileInfo $current) {
 
                 continue;
             }
-            if ('php' !== $file->getExtension() || !class_exists($className = $psr4Prefix.'\\'.$file->getBasename('.php'))) {
```
Comment (dmitryuk) [inline src/Symfony/Component/Routing/Loader/Psr4DirectoryLoader.php:86]: What about, for example, interfaces or traits?

**Review by chalasr — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @derrabus.

**Comment by stof** [comment]
shouldn't this also skip enums ?

**Comment by derrabus** [comment]
Enums don't cause an issue as far as I can tell. Shall I add one to the fixtures just in case?

**Comment by derrabus** [comment]
#48177

**Comment by stof** [comment]
The fact that they don't have an issue is probably because the AnnotationClassLoader for the routing does not validate them (and so tries to load routes defined on them but it is unlikely to find some)

**Comment by derrabus** [comment]
Yes. Not sure what happens though if someone tried to add `Route` attributes to a enum method. 🙈

**Comment by stof** [comment]
Probably a WTF error when trying to instantiate the controller (unless that part takes care of giving a better error message). However, I'm fine with failing for such case, as there is no valid reason for that (while a route in an abstract class might make sense in some cases where you have a child class belong loaded as well)

**Pull request merged**